pkg/controller/statefulset/statefulset_predownload_image.go

```diff
@@ -0,0 +1,175 @@
+	apps "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	intstrutil "k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/controller/history"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (dss *defaultStatefulSetControl) createImagePullJobsForInPlaceUpdate(sts *appsv1beta1.StatefulSet, currentRevision, updateRevision *apps.ControllerRevision) error {
+	if _, ok := updateRevision.Labels[appsv1alpha1.ImagePreDownloadCreatedKey]; ok {
+		return nil
+	} else if _, ok := updateRevision.Labels[appsv1alpha1.ImagePreDownloadIgnoredKey]; ok {
+		return nil
+	}
+
+	// ignore if update type is ReCreate
+	if string(sts.Spec.UpdateStrategy.Type) == string(appsv1beta1.RecreatePodUpdateStrategyType) {
+		klog.V(4).Infof("Statefulset %s/%s skipped to create ImagePullJob for update type is %s",
+			sts.Namespace, sts.Name, sts.Spec.UpdateStrategy.Type)
+		return dss.patchControllerRevisionLabels(updateRevision, appsv1alpha1.ImagePreDownloadIgnoredKey, "true")
+	}
+
+	// ignore if replicas <= minimumReplicasToPreDownloadImage
+	if *sts.Spec.Replicas <= minimumReplicasToPreDownloadImage {
+		klog.V(4).Infof("Statefulset %s/%s skipped to create ImagePullJob for replicas %d <= %d",
+			sts.Namespace, sts.Name, *sts.Spec.Replicas, minimumReplicasToPreDownloadImage)
+		return dss.patchControllerRevisionLabels(updateRevision, appsv1alpha1.ImagePreDownloadIgnoredKey, "true")
+	}
+
+	// ignore if all Pods update in one batch
+	var partition, maxUnavailable int
+	if sts.Spec.UpdateStrategy.RollingUpdate.Partition != nil {
+		// partition, _ = intstrutil.GetValueFromIntOrPercent(sts.Spec.UpdateStrategy.Partition, int(*sts.Spec.Replicas), true)
+		partition = int(*sts.Spec.UpdateStrategy.RollingUpdate.Partition)
+	}
+	maxUnavailable, _ = intstrutil.GetValueFromIntOrPercent(
+		intstrutil.ValueOrDefault(sts.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable, intstrutil.FromString("1")), int(*sts.Spec.Replicas), false)
+	if partition == 0 && maxUnavailable >= int(*sts.Spec.Replicas) {
+		klog.V(4).Infof("Statefulset %s/%s skipped to create ImagePullJob for all Pods update in one batch, replicas=%d, partition=%d, maxUnavailable=%d",
+			sts.Namespace, sts.Name, *sts.Spec.Replicas, partition, maxUnavailable)
+		return dss.patchControllerRevisionLabels(updateRevision, appsv1alpha1.ImagePreDownloadIgnoredKey, "true")
+	}
+
+	// opt is update option, this sectoin is to get update option
+	opts := &inplaceupdate.UpdateOptions{}
+	if sts.Spec.UpdateStrategy.RollingUpdate.InPlaceUpdateStrategy != nil {
+		opts.GracePeriodSeconds = sts.Spec.UpdateStrategy.RollingUpdate.InPlaceUpdateStrategy.GracePeriodSeconds
+	}
+
+	// ignore if this revision can not update in-place
+	inplaceControl := inplaceupdate.New(statefulsetClient, clonesetutils.RevisionAdapterImpl) // this name and method is actually generic
```

Comment:
use `revisionadapter.NewDefaultImpl()` instead of `clonesetutils.RevisionAdapterImpl`